modules/thermal_hydraulics/include/userobjects/NumericalFlux1D.h

```diff
@@ -0,0 +1,161 @@
+                                                const RealVectorValue & nLR,
+                                                const RealVectorValue & t1,
+                                                const RealVectorValue & t2) const;
+
+  /**
+   * Calculates the 3D flux vectors given "left" and "right" states
+   *
+   * This function is called only if the values are not already cached.
+   *
+   * @param[in] UL_3d  Vector of 3D flux inputs on the "left"
+   * @param[in] UR_3d  Vector of 3D flux inputs on the "right"
+   * @param[in] nLR   Direction from "left" to "right"
+   * @param[in] t1    1st tangent direction
+   * @param[in] t2    2nd tangent direction
+   * @param[out] FL   Flux vector to be added to "left" side
+   * @param[out] FR   Flux vector to be added to "right" side
+   */
+  virtual void calcFlux(const std::vector<ADReal> & UL_3d,
+                        const std::vector<ADReal> & UR_3d,
+                        const RealVectorValue & nLR,
+                        const RealVectorValue & t1,
+                        const RealVectorValue & t2,
+                        std::vector<ADReal> & FL,
+                        std::vector<ADReal> & FR) const = 0;
+
+  /**
+   * Returns the index of the region last entered
+   *
+   * Here "region" refers to a code path taken. For some fluxes, such as centered
+   * fluxes, there is just a single code path, but for others, such as those
+   * using an approximate Riemann solver, there are multiple. Riemann solvers
+   * have "regions" defined by the characteristic waves.
+   */
+  unsigned int getLastRegionIndex() const { return _last_region_index; }
+
+  /**
+   * Returns the total possible number of regions
+   *
+   * Here "region" refers to a code path taken. For some fluxes, such as centered
+   * fluxes, there is just a single code path, but for others, such as those
+   * using an approximate Riemann solver, there are multiple. Riemann solvers
+   * have "regions" defined by the characteristic waves.
+   */
+  virtual unsigned int getNumberOfRegions() const = 0;
+
+protected:
+  /**
+   * Converts a 1D flux input vector to a 3D flux input vector
+   *
+   * @param[in] U_1d  1D flux input vector
+   */
+  virtual std::vector<ADReal> convert1DInputTo3D(const std::vector<ADReal> & U_1d) const = 0;
+
+  /**
+   * Converts a 3D flux vector to a 1D flux vector
+   *
+   * @param[in] F_3d  3D flux vector
+   */
+  virtual std::vector<ADReal> convert3DFluxTo1D(const std::vector<ADReal> & F_3d) const = 0;
+
+  /**
+   * Applies direction transformation to a 3D flux vector
+   *
+   * @param[inout] F_3d  3D flux vector
+   * @param[in] nLR_dot_d   Dot product of direction from "left" to "right" with
+   *                        the flow channel direction
+   */
+  virtual void transform3DFluxDirection(std::vector<ADReal> & F_3d,
+                                        const ADReal & nLR_dot_d) const = 0;
```

Comment:
why AD for the direction?

Reply:
No good reason, probably oversight. I'll try changing.